### PR TITLE
[SPARK-41151][FOLLOW-UP][SQL] Keep built-in file _metadata fields nullable value consistent

### PR DESCRIPTION
**Pull request by Yaohua628**
### What changes were proposed in this pull request?
A follow-up PR of https://github.com/apache/spark/pull/38683. 

Apart from making `_metadata` struct not nullable, we should also make all fields inside of `_metadata` not nullable (`file_path`, `file_name`, `file_modification_time`, `file_size`, `row_index`).

### Why are the changes needed?
Consistent nullability behavior for everything

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
New UTs

### CODE REVIEW

**Comment by Yaohua628** [comment]
@cloud-fan @dongjoon-hyun @HeartSaVioR Sorry for the back and forth. 

[The previous PR](https://github.com/apache/spark/pull/38683), we changed the `_metadata` to not null. And I just realized we probably should make all fields inside of the `_metadata` (`file_path`, `file_name`, `file_modification_time`, `file_size`, `row_index`) not null as well for consistency.

Please let me know WDYT. As @cloud-fan mentioned, it should be fine to write not-null data into a nullable column. But my only concern is this change might break the existing stateful streaming schema compatibility check?

Also, cc @ala to confirm `row_index` will always be not null for supported file formats (e.g. Parquet)

Thanks for all your help!

**Comment by HeartSaVioR** [comment]
state store schema checker handles the compatibility for nullability. It does not only allow equality, but also allow the case when column for existing schema is nullable whereas column for new schema is non-nullable. So ensuring columns to be non-nullable would be OK for compatibility point of view.

**Review by HeartSaVioR — APPROVED** [review]
+1

**Comment by HeartSaVioR** [comment]
Let me see if there are further review comments today. I will merge this tomorrow if there is no outstanding comment.

**Comment by Yaohua628** [comment]
Thank you, Jungtaek! Also wanna confirm with @ala on nullability of `row_index`

**Comment by HeartSaVioR** [comment]
Ah OK, let's wait for feedback from @ala and ensure we make clear before merging it.

**Review by cloud-fan — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

According to the above discussion, gentle ping, @ala .

**Comment by ala** [comment]
Sorry, I was on PTO/sick for a couple of days. 

My idea was not to include the row index in `_metadata` for formats that cannot generate it. While we have _many_ Parquet readers, I believe all of them support row index generation right now, so that's not a problem.

For me the main concern might be that we might want to keep growing the `_metadata` struct (most recently: row id, cc @tomvanbussel, @juliuszsompolski). These new fields might not be immediately supported in all the readers (or if they are Databricks-internal, we might not want to support them in OSS readers at all). So while I think it's OK to make row index non-nullable, it would be an issue if we wanted to require that all `_metadata` fields are non-nullable in the future.

**Comment by HeartSaVioR** [comment]
@Yaohua628 Could you please push a new empty commit or rebase to master branch to retrigger build? Let's make sure build is green before merging this.

**Comment by Yaohua628** [comment]
@HeartSaVioR @ala some tests in `FileMetadataStructRowIndexSuite` are failed complaining: 
```
java.io.IOException: Required column is missing in data file. Col: [_tmp_metadata_row_index]
[info] 	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.checkColumn(VectorizedParquetRecordReader.java:375)
[info] 	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.initializeInternal(VectorizedParquetRecordReader.java:349)
[info] 	at org.apache.spark.sql.execution.datasources.parquet.VectorizedParquetRecordReader.initialize(VectorizedParquetRecordReader.java:181)
```

I had a fix [here](https://github.com/apache/spark/pull/38777/commits/0f98dd97a3c84a10be6fd7f75468e5d62f17fe82) to resolve failures (basically: keep the internal `_tmp_metadata_row_index` nullable, but `_metadata.row_index` is still not null). I don't fully understand what happened internally, could you take a look? Thanks!

**Comment by HeartSaVioR** [comment]
I don't have context for that, sorry. I'm OK either way if the nullability of column is guaranteed to not fluctuate.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:238]: unnecessary change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:276]: can we add a code comment to explain this nullability change?

**Comment by Yaohua628** [comment]
Addressed comments, thanks for taking a look!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:278]: shall we update `fileFormatReaderGeneratedMetadataColumns` to set nullablity as false?

Comment (Yaohua628) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala:278]: Thanks, Wenchen, I tried that before, but it failed many test cases in `FileMetadataStructRowIndexSuite`. See [this fix commit](https://github.com/apache/spark/pull/38777/commits/0f98dd97a3c84a10be6fd7f75468e5d62f17fe82) and [this comment](https://github.com/apache/spark/pull/38777#issuecomment-1332079823).

I don't have much context on the `row_index`, not sure what caused the issue, any idea? Thanks!

**Comment by ala** [comment]
Well, the issue seems to be that  the vectorized reader recognizes the row index column as a "missing column" (aka. columns that are not read from the file, but instead populated by a higher layer in the reader). Since these are normally populated with nulls, it's a problem if the data type is non-nullable. https://github.com/apache/spark/blob/0f1c515179e5ed34aca27c51f500c26ca19cc748/sql/core/src/main/java/org/apache/spark/sql/execution/datasources/parquet/VectorizedParquetRecordReader.java#L372-L376
We could tweak this `if` condition to not throw on generate column/row index, or use the workaround you put in place already.

**Comment by Yaohua628** [comment]
Thanks for the explanation, @ala! I am OK either way cc @cloud-fan @HeartSaVioR feel free to merge it if you think it is OK ^

**Comment by HeartSaVioR** [comment]
It seems OK to me as well but I'll lean on @cloud-fan on the decision as I'm not an expert on this subject.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.3!

**Comment by cloud-fan** [comment]
oh it conflicts with 3.3, @Yaohua628 can you open a backport PR? thanks!

**Pull request closed without merging**
